tls/s2n_cipher_suites.c

```diff
@@ -219,6 +219,29 @@ struct s2n_cipher_suite s2n_null_cipher_suite = {
     .record_alg = &s2n_record_alg_null,
 };
 
+/* Both the group start and group end cipher suites are never negotiated. To ensure
+ * this, the IANA value is set to TLS_NULL_WITH_NULL_NULL. s2n checks that the negotiated
+ * cipher IANA id is never this value (see s2n_connection_get_cipher_iana_value) therefore preventing
+ * these ciphers from being negotiated.
+ */
+struct s2n_cipher_suite s2n_equal_preference_group_start = {
+    .available = 0,
```

Comment:
Clarifications:
1. Isn't it impossible for these ciphers to be set by the client: https://github.com/aws/s2n-tls/blob/main/tls/s2n_client_hello.c#L603?
2. Similarly, the server can't negotiate these ciphers: https://github.com/aws/s2n-tls/pull/3488/files#diff-d0b6d9d296ee7b71a99c83ad1423142d1cc7d6b62a84ebfc31a9b0f633a4f323R1297. As to proving that this is 'never chosen by the server', I can add some unit tests that uses a preference list of just delimiters. But this doesn't really 'prove' the invariant... I think I would need formal verification for that.
3. Would a simple test asserting that `s2n_equal_preference_group_start->available == 0` be sufficient?

Reply:
Both cases should definitely be impossible, which is why we can test that they're impossible.

Checking s2n_equal_preference_group_start->available == 0 would probably be sufficient if tests already exist to verify that we don't send or choose unavailable cipher suites (and I believe we do test those cases). But if you're really paranoid, you can also write more specific tests to make very sure we never send or choose these cipher suites, since they're a special case.